### PR TITLE
Bug 1241464 - Create a eslint formatter that can filter out known err…

### PR DESCRIPTION
**Pull request by julienw**
…ors in a file

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9450ed4678fedd284fb58826f40a5e78d01abe06

File: build/eslint/xfail.list [hunk 1]
```diff
@@ -1,298 +1,206 @@
...
+apps/communications/contacts/test/unit/views/multiple_select_test.js:22:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/communications/contacts/test/unit/views/settings_test.js:134:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/communications/contacts/test/unit/views/settings_test.js:474:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/costcontrol/js/settings/settings.js:170:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/costcontrol/test/unit/formatting_test.js:46:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/costcontrol/test/unit/formatting_test.js:228:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/email/build/email.build.js:3:10: Parsing error: Unexpected token [Error]
+apps/email/build/gelam_worker.build.js:3:10: Parsing error: Unexpected token [Error]
+apps/email/build/main-frame-setup.build.js:3:6: Parsing error: Unexpected token [Error]
```
Comment (julienw) [inline build/eslint/xfail.list]: not sure why these files do not parse properly

File: build/eslint/xfail.list [hunk 2]
```diff
@@ -1,298 +1,206 @@
...
+apps/keyboard/build/settings-config.js:72:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/js/imes/jspinyin/tests/init.js:103:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/js/imes/jszhuyin/lib/jszhuyin_server.js:42:7: Using postMessage() with wildcard targets is not allowed. [Error/no-wildcard-postmessage/no-wildcard-postmessage]
+apps/keyboard/js/views/emoji_key_view.js:22:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/js/views/key_view.js:118:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/js/views/key_view.js:127:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/js/views/key_view.js:134:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/js/views/key_view.js:139:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/keyboard/test/unit/keyboard/layout_normalizer_test.js:64:30: Parsing error: Unexpected token [Error]
```
Comment (julienw) [inline build/eslint/xfail.list]: this is the array comprehension syntax; I don't know if a newer babel could parse it ?

File: build/eslint/xfail.list [hunk 3]
```diff
@@ -1,298 +1,206 @@
...
+apps/music/components/gaia-header/dist/gaia-header-es5.js:215:15: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/music/components/gaia-header/dist/gaia-header-es5.js:501:13: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/music/components/gaia-header/dist/gaia-header.js:180:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/music/components/gaia-header/dist/gaia-header.js:454:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/music/components/gaia-theme/lib/gaia-theme-selector.js:12:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/music/components/gaia-toolbar/gaia-toolbar.js:26:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/music/components/poplar/poplar.js:264:3: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/ringtones/js/tone_list.js:51:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/settings/js/modules/addon_manager.js:246:12: Parsing error: Unexpected token [Error]
```
Comment (julienw) [inline build/eslint/xfail.list]: this is the "for...of"  construct; pretty sure babel could parse this.

File: build/eslint/xfail.list [hunk 4]
```diff
@@ -1,298 +1,206 @@
...
+apps/sms/views/conversation/js/recipients.js:509:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/conversation/test/unit/conversation_test.js:283:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/conversation/test/unit/conversation_test.js:4452:9: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/conversation/test/unit/conversation_test.js:6320:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/conversation/test/unit/subject_composer_test.js:138:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/inbox/js/inbox.js:789:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/inbox/test/unit/inbox_test.js:101:5: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/shared/js/contact_renderer.js:275:7: Unsafe assignment to innerHTML [Error/no-unsafe-innerhtml/no-unsafe-innerhtml]
+apps/sms/views/shared/tes
```
Comment (julienw) [inline build/eslint/xfail.list]: array comprehension again

File: build/eslint/gaia-formatter.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  var xfail;
+  try {
+    xfail = require('fs').readFileSync(xfailFilename, 'utf-8').split('\n');
+  } catch(e) {
+    debug(`Error while loading ${xfailFilename}`, e);
+    return {};
+  }
+
+  // xfail list uses the format output by "eslint -f unix"
+  const reParseXfailLine = /^([^:]+):(\d+:\d+):(.+)\[[^\/]+\/(.+)\]$/;
```
Comment (julienw) [inline build/eslint/gaia-formatter.js]: this doesn't handle properly files that file to parse

File: build/eslint/gaia-formatter.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  return xfail.reduce((result, cur) => {
+    const parsed = reParseXfailLine.exec(cur);
+    if (!parsed) {
+      return result;
+    }
+
+    const filename = parsed[1].replace(reStartWithPwd, '');
+    const errorLocation = parsed[2];
+    const errorMessage = parsed[3].trim();
+    const errorCode = parsed[4];
```
Comment (julienw) [inline build/eslint/gaia-formatter.js]: I don't have destructuring assignments in node :(

File: build/eslint/gaia-formatter.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
+const debug = require('debug')('eslint:gaia-formatter');
+
+const xfailFilename = 'build/eslint/xfail.list';
+const pwd = process.cwd().replace(/\/$/, '');
```
Comment (julienw) [inline build/eslint/gaia-formatter.js]: I need to make sure we run this in the gaia root directory

**Comment by julienw** [comment]
eslint -f <path to formatter> <paths>

File: build/eslint/xfail.list (hunk 1 above)
Comment (mozfreddyb) [inline build/eslint/xfail.list]: I'm surprised this eslint policy change introduces new innerHTML violation. Why?

Most of the files in email/build/ are actually json, not js. Maybe we should file a follow-up bug to rename them to json?

File: build/eslint/xfail.list (hunk 2 above)
Comment (mozfreddyb) [inline build/eslint/xfail.list]: There were other things with array (and object?) comprehension syntax that I managed to make *not fail* by changing the parser to babel-eslint.
In any case, I'm +1 on bumping up the babel version

**Pull request closed without merging**